compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java

```diff
@@ -982,6 +989,30 @@ void handleDeclaredVarInForeach(BLangVariable variable, BType rhsType, SymbolEnv
         }
     }
 
+    private BType getInferredTypeForFinalVar(BLangExpression expr, BType originalType) {
+        // Inferred type for final variables declared with var should be the precise type
+        switch (expr.getKind()) {
+            case SIMPLE_VARIABLE_REF:
+                BLangSimpleVarRef ref = (BLangSimpleVarRef) expr;
+                if (ref.symbol.getKind() == SymbolKind.CONSTANT) {
+                    return ((BConstantSymbol) ref.symbol).type;
+                }
+                return originalType;
+            case NUMERIC_LITERAL:
+            case LITERAL:
+                return new BFiniteType(createTypeSymbol(SymTag.FINITE_TYPE), Collections.singleton(expr));
+            case LIST_CONSTRUCTOR_EXPR:
```

Comment:
Can we mention in https://github.com/ballerina-platform/ballerina-lang/issues/22959 that we need to add support for mapping constructor expressions here too?